internal/auth/authenticator_oidc.go

```diff
@@ -328,14 +341,47 @@ func (oa *OIDCAuthenticator) Authenticate(msg *message.Message) (bool, []action.
 
 	// Verify the cookie to make sure the user is authenticated
 	if oauthArgs.cookie != "" {
-		err := oa.checkCookie(oauthArgs.cookie, extractDomainFromHost(oauthArgs.host))
+		idToken, err := oa.decryptCookie(oauthArgs.cookie, domain)
 		if err != nil {
+			// CoreOS/go-oidc does not have error types, so the errors are handled using strings
+			// comparison.
+			if errors.Is(err, &oidc.TokenExpiredError{}) || strings.Contains(err.Error(), "oidc:") {
+				authorizationURL, e := oa.builaAuthorizationURL(domain, oauthArgs)
+				if e != nil {
+					return false, nil, e
+				}
+
+				logrus.Infof("Authentication failed, redirecting to OIDC provider %s, reason: %s", authorizationURL, err)
+
+				return false, []action.Action{BuildRedirectURLMessage(authorizationURL)}, nil
+			}
+
 			return false, nil, err
-		} else {
+		}
+
+		if len(oauthArgs.tokenClaims) == 0 {
 			return true, nil, nil
+		} else {
+			// Extract token claims.
+			actions, err := BuildTokenClaimsMessage(idToken, oauthArgs.tokenClaims)
+			if err != nil {
+				return false, nil, err
+			}
+
+			return true, actions, nil
 		}
+
 	}
 
+	authorizationURL, err := oa.builaAuthorizationURL(domain, oauthArgs)
+	if err != nil {
+		return false, nil, err
+	}
+
+	return false, []action.Action{BuildRedirectURLMessage(authorizationURL)}, nil
+}
+
+func (oa *OIDCAuthenticator) builaAuthorizationURL(domain string, oauthArgs OAuthArgs) (string, error) {
```

Comment:
typo: rename to `buildAuthorizationURL`

Reply:
Thank you for noticing it!

Reply:
The tests don't pass, it just hang. Already retriggered twice, same result.

Retrying another time :crossed_fingers:

Reply:
Test passed - I wonder if this is not linked to the HAProxy version - running in the CI, I see version 2.9.9 - and when running locally, I get v2.9.7.

Anyway, not linked to this patch - I'm merging it.